coderd/httpapi/status_writer.go

```diff
@@ -0,0 +1,69 @@
+package httpapi
+
+import (
+	"bufio"
+	"net"
+	"net/http"
+
+	"golang.org/x/xerrors"
+)
+
+var _ http.ResponseWriter = (*StatusWriter)(nil)
+var _ http.Hijacker = (*StatusWriter)(nil)
+
+// StatusWriter intercepts the status of the request and the response body up
+// to maxBodySize if Status >= 400. It is guaranteed to be the ResponseWriter
+// directly downstream from Middleware.
+type StatusWriter struct {
```

Comment:
Suggestion (non-blocking): I think it could be beneficial to make this a package-private type and expose the fields via functions.

Benefit is that a caller can't access the underlying ResponseWriter (bypassing StatusWriter), but can still test for the values via `if s, ok := rw.(interface{Status() int}); ok { s.Status() }`

Reply:
I'm not sure I'm following. Do you just want to create a `StatusWriter` interface and unexport the current struct? Alternatively we could simply unexport the `ResponseWriter` on the struct.

Reply:
Ah, sorry it was unclear, I was suggesting two changes:

1. Unexport `ResponseWriter` => `responseWriter`
2. Change fields from `Status` => `func(*statusWriter) Status() int`

Number 1 was the main thing (to protect from direct ResponseWriter access), number 2 was a little extra which would allow the values to be inspected outside the package, if the need ever arises. But not necessary by any means.

In the case of 2, my example of `if s, ok := rw.(interface{Status() int}); ok { s.Status() }` was hinting at that we don't need to export anything from this package. If that need ever arises (inspect value outside package), it could be done with an inline interface.

I could've omitted that part though, I don't think it will be needed. 😄